ddl/cluster.go

```diff
@@ -491,15 +494,73 @@ func finishFlashbackCluster(w *worker, job *model.Job) error {
 					return err
 				}
 			}
+			if job.IsDone() || job.IsSynced() {
+				gcSafePoint, err := gcutil.GetGCSafePoint(sess)
+				if err != nil {
+					return err
+				}
+				if err = UpdateFlashbackHistoryTSRanges(t, flashbackTS, t.StartTS, gcSafePoint); err != nil {
+					return err
+				}
+			}
 			if err = t.SetFlashbackClusterJobID(0); err != nil {
 				return err
 			}
 		}
 		return nil
 	})
 	if err != nil {
+		logutil.BgLogger().Warn("Finish flashback cluster meets error", zap.Error(err))
 		return err
 	}
 
 	return nil
 }
+
+// CheckFlashbackHistoryTSRange checks flashbackTS overlapped with history time ranges or not.
+func CheckFlashbackHistoryTSRange(m *meta.Meta, flashbackTS uint64) error {
+	tsRanges, err := m.GetFlashbackHistoryTSRange()
+	if err != nil {
+		return err
+	}
+	for _, tsRange := range tsRanges {
+		if tsRange.StartTS <= flashbackTS && flashbackTS <= tsRange.EndTS {
+			return errors.Errorf("FlashbackTs overlapped, old range: [%s, %s], flashbackTS: %s",
+				oracle.GetTimeFromTS(tsRange.StartTS), oracle.GetTimeFromTS(tsRange.EndTS), oracle.GetTimeFromTS(flashbackTS))
+		}
+	}
+	return nil
+}
+
+// UpdateFlashbackHistoryTSRanges insert [startTS, endTS] into FlashbackHistoryTSRange.
+func UpdateFlashbackHistoryTSRanges(m *meta.Meta, startTS uint64, endTS uint64, gcSafePoint uint64) error {
+	tsRanges, err := m.GetFlashbackHistoryTSRange()
+	if err != nil {
+		return err
+	}
+	if len(tsRanges) != 0 && tsRanges[len(tsRanges)-1].EndTS >= endTS {
+		return errors.Errorf("Invalid flashback ts range, last flashback end time: %s, now: %s",
+			oracle.GetTimeFromTS(tsRanges[len(tsRanges)-1].EndTS), oracle.GetTimeFromTS(endTS))
+	}
+
+	newTsRange := make([]meta.TSRange, 0, len(tsRanges))
+
+	for _, tsRange := range tsRanges {
+		if tsRange.EndTS < gcSafePoint {
+			continue
+		}
+		if startTS > tsRange.EndTS {
+			newTsRange = append(newTsRange, tsRange)
+		} else if startTS < tsRange.StartTS {
+			newTsRange = append(newTsRange, meta.TSRange{StartTS: startTS, EndTS: endTS})
+			break
+		} else {
+			// If startTS in range [tsRange.StartTs, tsRange.EndTs], it's an impossible startTS.
+			return errors.Errorf("Invalid flashback ts range, startTS in old time range")
```

Comment:
Should not happend since `flashback` is cluster-wide locking command, this comment should be more friendly.

Reply:
I mean the error, actually, comment will not be logged, but the error will.

Reply:
In fact, I mean the error, which may be logged or returned.